File: wpilibc/src/main/native/cpp/simulation/AnalogEncoderSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/simulation/AnalogEncoderSim.hpp"

#include "wpi/hardware/rotation/AnalogEncoder.hpp"
#include "wpi/simulation/SimDeviceSim.hpp"

using namespace wpi::sim;

AnalogEncoderSim::AnalogEncoderSim(const wpi::AnalogEncoder& encoder) {
  wpi::sim::SimDeviceSim deviceSim{"AnalogEncoder", encoder.GetChannel()};
  m_positionSim = deviceSim.GetDouble("Position");
}

void AnalogEncoderSim::Set(double value) {
  m_positionSim.Set(value);
}

double AnalogEncoderSim::Get() {
  return m_positionSim.Get();
}

File: wpilibj/src/main/java/edu/wpi/first/wpilibj/simulation/AnalogEncoderSim.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.simulation;

import edu.wpi.first.hal.SimDouble;
import edu.wpi.first.wpilibj.AnalogEncoder;

/** Class to control a simulated analog encoder. */
public class AnalogEncoderSim {
  private final SimDouble m_simPosition;

  /**
   * Constructs from an AnalogEncoder object.
   *
   * @param encoder AnalogEncoder to simulate
   */
  public AnalogEncoderSim(AnalogEncoder encoder) {
    SimDeviceSim wrappedSimDevice =
        new SimDeviceSim("AnalogEncoder" + "[" + encoder.getChannel() + "]");
    m_simPosition = wrappedSimDevice.getDouble("Position");
  }

  /**
   * Set the position.
   *
   * @param value The position.
   */
  public void set(double value) {
    m_simPosition.set(value);
  }

  /**
   * Get the simulated position.
   *
   * @return The simulated position.
   */
  public double get() {
    return m_simPosition.get();
  }
}
